### ISSUE TITLE
Request for Comments: Magithub Manager

### ISSUE THREAD

**Issue by vermiculus** [vermiculus/magithub#72]
Some issues (like #67) would benefit from a dedicated mode to manage issues and pull requests. I've separately envisioned such a mode that would allow more specialized interaction with these objects beyond what is appropriate in `magit-status` (as there are only so many keys I can safely bind there), such as issue preview, org-mode integration, and yes, issue prioritization :wink:

~~I've got basic scaffolding up to list issues+pull requests in `tabulated-list-mode`, but I haven't implemented any features yet. With this interface in mind (or if you think of another), what kind of functionality would you like to see here?~~

After [discussion](https://github.com/vermiculus/magithub/issues/72#issuecomment-280212902), I've been sufficiently sold on [`sigma/gh.el`](https://github.com/sigma/gh.el) and will be moving to that platform before the next release. As such, feel free to suggest anything that comes to mind. [If the API supports it](https://developer.github.com/v3/), it looks like `gh.el` provides the means to get it done.

## ~~Links to Hub feature requests~~

* ~~Comment on issues and pull requests: github/hub#623 github/hub#1203~~
* ~~Create milestones: github/hub#1343~~
* ~~List notifications: github/hub/issues/1415~~
* ~~Pretty much everything else: github/hub#1416~~

**Comment by LefterisJP** [comment]
I would love it if I could manage the following from inside Emacs:

- Issues and labeling.
      - Creation of new labels and issues
      - Assigning labels and people to issues
      - Reading issues comments and replying to them
- Using the Github Reviewing system from inside Emacs. That would be really amazing (but I guess rather hard to implement)
- Milestone management.
     - Listing milestones and issues associated with those milestones.
     - Creating new milestones
     - Adding/removing issues to a milestone

This is just a wishlist, and I honestly have no idea how hard or trivial any of the above is to implement. I would love to help if I can.

**Comment by vermiculus** [comment]
Excellent suggestions, thank you!

> I would love to help if I can.

If you can find API for any or all of these, that would be a great help :) For example, [creating comments is supported by the API](https://developer.github.com/v3/issues/comments/#create-a-comment), but `hub` doesn't support it yet. I can lobby for support on that repository, though.

**Comment by rgrinberg** [comment]
Another github feature that I would love if was available is viewing your github notifications inside Emacs. I would love to be able to see recently changed issues/prs in repos that I'm watching.

Also the github frontpage that shows you a recent activity of people that you're following would be a nice thing to have as well. When I used vim, this plugin covered that for me https://github.com/junegunn/vim-github-dashboard Would be great to restore that in Emacs.

**Comment by blaenk** [comment]
Have you looked at using [gh.el](https://github.com/sigma/gh.el) instead of being limited to whatever hub supports? This would even help with removing the dependency on hub in the long term, if that's a desirable goal. Some of the functionality can probably be easily replicated in elisp with the help of magit's git API, such as checking out pull requests to a local branch. It's probably not as complicated as it may seem. The bulk of what hub provides seems to be parsing command arguments and translating them into the desired sequence of regular git commands.

I realize you're just scaffolding/prototyping at the moment, but I personally am not a fan of tabulated-list-mode, it feels too static and restrictive. I wonder if you can leverage the infrastructure magit uses for it's magit-status screen, considering they do expose the magit-popup for external use? That would be amazing because we could easily leverage the ability to collapse entire sections or levels of sections, such as collapse all pull requests, issues, etc., as well as the ability to jump to a particular section, as well as existing syntax highlighting definitions that already exist in many themes for magit-status.

I agree with creating a separate magithub screen for the reasons you state, such as to be able to reclaim and use binds that are otherwise taken in magit-status. If anything I think it would be nice, however, to add bridges for functionality where appropriate. At the very least, perhaps show stubs in magit-status like `Pull Requests: 45` where if one `ENTER`s on it, it'll take you to the magithub screen's section for pull requests. That way we can still get an at-a-glance heads-up on the github state and an easy way to jump into it from the arguably more frequently used magit-status.

It may even be useful in some cases to nest deeper, so that magithub-status would show an overview of pull requests for example and when one ENTERs into it, it'd show a magithub-pull-requests-status screen that's again a screen similar to magit-status and magithub-status but is specific to pull requests.

Also if you retain the magit-popup interface to mirror magit-status' that would be amazing. As I mentioned above, IIRC magit-popup is made available as a library for use "outside" of magit.

TL;DR I agree with a separate, dedicated "magithub-status" type screen if it mirrors magit-status as much as possible (e.g. appearance and functionality, sections with headings, each collapsible) preferably using the same infrastructure magit uses for its magit-status screen instead of reinventing the wheel, and also uses magit-popup, and provides bridges between the two where appropriate. The reason for this would be to minimize the friction incurred by context-switching between the two and to reinforce/retain the user interaction language.

**Comment by blaenk** [comment]
So as I mentioned, magit-popup can definitely be used outside of magit AFAIK. See the [documentation](https://magit.vc/manual/magit-popup.html).

As for [magit-status](https://github.com/magit/magit/blob/master/lisp/magit-status.el), it seems like they define their own magit-status-mode, which is useful for being able to define a keymap specific to that mode among other things, and then they use generic functions such as [`magit-insert-section`](https://github.com/magit/magit/blob/3cbf660eb5a130d01d1ca1a697435b0b8c88f9cb/lisp/magit-section.el#L643) to insert sections, [`magit-insert-heading`](https://github.com/magit/magit/blob/3cbf660eb5a130d01d1ca1a697435b0b8c88f9cb/lisp/magit-section.el#L751) to insert headings for sections, then there's function [`magit-define-section-jumper`](https://github.com/magit/magit/blob/3cbf660eb5a130d01d1ca1a697435b0b8c88f9cb/lisp/magit-section.el#L280) to define a key that can jump to a particular section, and so on.

Documentation for this and other "plumbing" APIs is available [in the plumbing section](https://magit.vc/manual/magit.html#Plumbing), in particular the section on [section plumbing](https://magit.vc/manual/magit.html#Section-Plumbing). Notice that there's also a section on [git plumbing](https://magit.vc/manual/magit.html#Calling-Git) which may be useful in conjunction with gh.el in eventually removing the dependency on hub (if that's something that should be pursued).

**Comment by vermiculus** [comment]
> Have you looked at using gh.el instead of being limited to whatever hub supports?

I have, but I really want to stay away from the auth flow required. When I least looked at it, `gh.el` wasn't really viable, but perhaps I'll take another look when we've run the `hub` train dry. For now, it's a useful abstraction to keep things simple; I've tried to structure Magithub code to swap it out if that's ever necessary, though 😉

Re `tabulated-list-mode` vs. `magit-insert-section`: *yes*!! I was thinking about this over lunch this afternoon and I really want to get home to see what I can figure out 😄 What made me consider it is the restrictions of tabulated-list-mode for separating issues from pull requests. I'll update with screenshots sometime soon.

**Comment by blaenk** [comment]
That's great to hear.

I'm curious what auth flow you're concerned about with respect to gh.el? For example, I use gist.el which uses gh.el, it did require me to make an oauth token which I then set in my gitconfig (this doesn't have to be in ~/.gitconfig itself, instead I put mine in a ~/.gitconfig.local that I don't check in with my dotfiles by using the [include directive](https://github.com/blaenk/dots/blob/master/git/.gitconfig#L112-L113)) and which can now be used by other tools that may need it. I think it would be possible to automate this though like hub itself does, where it asks for user & pass the first time then uses that to retrieve an oauth token and then saves that oauth token. In fact it seems like gh.el may have some [support for this](https://github.com/sigma/gh.el/blob/master/gh-auth.el).

I sympathize with the aversion to the one-time inconvenience of having to get an oauth token (if we don't do something like hub does, which seems straightforward), but I feel like it's an inconvenience outweighed by being able to do more things, the "right way" rather than having to wait and petition hub developers for new features/changes and hope they're receptive and accommodating. In the worst case, we could even just parse ~/.config/hub to read out the oauth token that hub retrieved and use that.

To repeat, I'm not suggesting we purge the hub dependency outright right away, but I do think we should be augmenting functionality through gh.el and magit plumbing APIs rather than limiting ourselves to the extent of hub's functionality.

**Comment by blaenk** [comment]
I can confirm. I just now ran with edebug [`gh-auth-get-oauth-token`](https://github.com/sigma/gh.el/blob/master/gh-auth.el#L78) with my oauth-token commented out in my config file. It proceeded to ask me for my username and password, then my 2fa code, then it retrieved an oauth token from github and automatically saved it in my gitconfig. It was completely painless. Is this the flow you have a problem with?

If the last step (saving it in the config file) is not desirable then the function can simply be modified to not include the `(gh-auth-remember profile :token token)`, then it can be handled some other way, either notifying the user that they should save it themselves or saving it in another file like hub does in ~/.config/hub, like somewhere in the ~/.emacs.d.

In fact, it will probably be desirable to appropriate that function and customize it, for example the sexp `(funcall 'gh-oauth-auth-new api '(user repo gist))` can be modified to specify alternative scopes (the final parameter is a list of scopes that the token should have permission for), in order to only request scopes for features that magithub supports, for example.

Here's the function in its entirety:

``` elisp
(defun gh-auth-get-oauth-token ()
  (let* ((profile (gh-profile-current-profile))
         (token (or (plist-get (cdr (assoc profile gh-auth-alist)) :token)
                    (plist-get (cdr (assoc profile gh-profile-alist)) :token)
                    (gh-config "oauth-token"))))
    (when (not token)
      (let* ((api (make-instance 'gh-oauth-api))
             (tok (and (fboundp 'gh-oauth-auth-new)
                       (oref (oref (funcall 'gh-oauth-auth-new api
                                            '(user repo gist)) :data)
                             :token))))
        (setq token (or tok (read-string "GitHub OAuth token: ")))
        (gh-set-config "oauth-token" token)))
    (gh-auth-remember profile :token token)
    token))
```

**Comment by vermiculus** [comment]
After trying out `gh.el` in a fresh VM, I must say I'm impressed with its default authentication workflow. I don't know what I tried those years ago, but it's clearly improved. Given that there's no client-side parsing, I've no problem moving over to it. It's still quite poorly documented, so any help here (i.e. upstream @ gh.el) would be appreciated; the last time `docs/` was updated was two years ago.

> the "right way"

This is a matter of opinion 😉 IMHO, petitioning upstream is more useful to the broader community (even if it hinders downstream projects at times). That said, there's a huge usability boon from a completely in-emacs experience.

> If the last step (saving it in the gitconfig file) is not desirable

Given that this doesn't even save your password, I think this is fine.

> it will probably be desirable to appropriate that function and customize it

If all we want to do is request more permissions, then it is more appropriate to make the choice available upstream.

----

With this in mind, feel free to suggest features to the fullest extent allowed by the API and under the assumption that we'll be using a magit-status-*like* mode instead a tabular list. I'll update the original post body.

**Comment by tarsius** [comment]
Good to see you working on this again! I think moving away from `hub` is a good idea. I believe I have already mentioned that earlier, and also that I still intend to add "collaboration features" to Magit itself.

I have gotten closer to actually getting started with that, now that I have opened magit/magit#2972. As you can see on that issue, I am currently more concerned with fundamental design decisions than with individual features. Never-the-less that issue contains a long but incomplete list of things I would like to implement eventually. I am happy to see that some of the features that I have in mind, are also being mentioned here, more verbosely.

Maybe the most important part of that issue as far as the future of `magithub` is concerned, is the following:

> I would also like to encourage the authors of these packages to continue their work. Some competition won't hurt. Of course you are also welcome to "rebase" your package onto Magit, once I am done with the ground work. Either as part of Magit or an extension. Or not at all, that is perfectly fine too.

I should go further than that and say that some **collaboration** won't hurt! That being said, I am currently still busy outlining my future work on Magit overall (which involves equally important changes such as improved performance and better abstractions). So right now I am not quite ready to collaborate just yet.

Please continue working on this. I won't provide much feedback at this point, but think you are going into the right direction. I have somewhat more far reaching goals for what I want to eventually find its way into Magit itself. But from what I have seen, some of that could come from here. Provided you want to do that.

**Comment by tarsius** [comment]
To comment on just a few points raised here:

As I said, I think it is a good idea to move away from the `hub` program to an all-elisp approach. I am not so sure `gh.el` is the right choice though. In the past I have contributed to it, but eventually moved away and instead written the much simpler `ghub.el`.

The main issue I had with `gh.el` is that it is to complex. It was the package that introduced my to eieio/clos, and for that I am very thankful, I use that extensively now. But I don't think that was the right choice for `gh.el`. Eieio was used with the goal of making things simpler, but my impression was that the opposite was the case; it made things much more complicated. In practice the complexity had three consequences:

1. I got back a response object with contained the data as another object, when I actually just wanted the data as a plain old alist. Also to make a request one had to create a request object, which in certain scenarios involved jumping through some hoops, when all I wanted to do was pass an api path (a string), and maybe some data (an alist).

2. When there was a bug, it was very hard to locate what method of what object was at fault. But maybe that was just me being new to clos/eieio.

3. Many connections failed. I never figured out whose fault that was, it might have been `url.el`'s or `gh.el`'s, or maybe Github was to blame. But since I have moved on to `ghub.el` I never had any such issues anymore (but maybe that's just because of the bad error handling).

So that's why I created `ghub.el` a very tiny abstraction. If there should be bug, then it should be easy to find out where it is, there is so little code. And all the fancy object stuff that I did not need is gone.

That being said, `ghub.el` is far from perfect. Most importantly it should have very good error handling, which it clearly has not right now. And it should probably provide a setup command that guides the user in acquiring and storing the access token.

I have opened an issue about using a common approach to storing the credential across all api libraries or package that do it by themselves:  magit/magit#2973.

The ideal outcome in my eyes would be if someone added the two features mentioned above to `ghub.el` (possibly porting things from `gh.el`), and then also changed `gh.el` to use `ghub.el` internally).

**Comment by blaenk** [comment]
I was not aware of `ghub.el` but I agree with the general idea that eioeio seemingly over complicates `gh.el`, makes it a bit tricky to figure out what's going on and where, an issue compounded by the lack of documentation.

> And it should probably provide a setup command that guides the user in acquiring and storing the access token.

Yeah this would be very useful, and IMO a core necessity for downstream packages. It's pretty nice how gh.el gracefully and automatically handles retrieving an oauth token if none is present in `git config`, asking for the user/pass in order to request it, also handling two factor authentication if present.

**Comment by IvanMalison** [comment]
> The ideal outcome in my eyes would be if someone added the two features mentioned above to ghub.el (possibly porting things from gh.el), and then also changed gh.el to use ghub.el internally).

I personally didn't find gh.el to be too hard to work with when I contributed to it when I wrote https://github.com/IvanMalison/github-search

I think that gh.el does a lot of good things. Though I will admit that I do sometimes get failed connections.

**Comment by vermiculus** [comment]
With the above issues now open, I'm going to close this issue. Feel free to have more conversation here (I'll read it), but I think we have some good direction now 😃
